### PR TITLE
trie: ensure the range is monotonic increasing

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

File: trie/proof.go [hunk 1]
```diff
@@ -364,102 +393,77 @@ func unset(parent node, child node, key []byte, pos int, removeLeft bool) error
...
+	if len(keys) == 0 {
+		return errors.New("empty proof"), false
+	}
+	// Ensure the received batch is monotonic increasing.
+	for i := 0; i < len(keys); i++ {
+		if i+1 == len(keys) {
+			break
+		}
+		if bytes.Compare(keys[i], keys[i+1]) >= 0 {
+			return errors.New("non-monotonic increasing range"), false
```
Comment (holiman) [inline trie/proof.go:419]:
```suggestion
			return errors.New("range is not monotonically increasing"), false
```

File: trie/proof.go [hunk 2]
```diff
@@ -408,7 +408,16 @@ func VerifyRangeProof(rootHash common.Hash, firstKey []byte, keys [][]byte, valu
 		return fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values)), false
 	}
 	if len(keys) == 0 {
-		return fmt.Errorf("empty proof"), false
+		return errors.New("empty proof"), false
+	}
+	// Ensure the received batch is monotonic increasing.
+	for i := 0; i < len(keys); i++ {
+		if i+1 == len(keys) {
```
Comment (karalabe) [inline trie/proof.go:415]: Why not `for i := 0; i < len(keys)-1; i++` instead of this extra break?

Comment (rjl493456442) [reply trie/proof.go:415]: If there is only 1 element, then it's out of bound.

**Comment by karalabe** [comment]
That's why -1 at the length

File: trie/proof.go (hunk 2 above)
Comment (rjl493456442) [reply trie/proof.go]: Fixed

**Review by Platinumwrist — COMMENTED** [review]
trie/proof.go trie/proof_test.go

**Pull request merged**